admin/controllers/pupil.js

```diff
@@ -95,27 +82,142 @@ const postAddPupil = async (req, res, next) => {
+  const stream = fs.createReadStream(uploadFile.file)
+  const csvStream = csv()
+    .on('data', (data) => { csvData.push(data) })
+    .on('end', async() => {
+      // Remove error column and headers from data
+      if (csvData.some(p => p[6])) csvData.map((r) => r.splice(6, 1))
+      let headers = csvData.shift(0)
+      // validate each pupil
+      const pupils = []
+      csvData = await Promise.all(csvData.map(async(p) => {
+        const pupil = new Pupil({
+          school: school._id,
+          upn: p[3],
+          foreName: p[0],
+          lastName: p[2],
+          middleNames: p[1],
+          gender: p[5],
+          dob: p[4],
+          pin: null,
+          pinExpired: false
+        })
+        try {
+          const dob = p[4].split('/')
+          const pupilData = Object.assign({
+            'dob-day': dob[1],
+            'dob-month': dob[0],
+            'dob-year': dob[2]
+          }, pupil._doc)
+          await validatePupil(pupil, pupilData)
+        } catch (err) {
+          p[6] = []
+          Object.keys(err.errors).forEach((e) => p[6].push(err.errors[e].replace(/ \\"/g, '')))
+          p[6] = p[6].join(', ')
+        }
+        pupils.push(pupil)
+        return p
+      }))
+      // Generate csv with errors
+      if (csvData.some(p => p[6])) {
+        const errorsCsv = []
+        headers.push('Errors')
+        errorsCsv.push(headers)
+        csvData.forEach((p) => errorsCsv.push(p))
+        const writeToString = promisify(csv.writeToString)
+        const cvsStr = await writeToString(errorsCsv, { headers: true })
+        // Upload csv to Azure
+        try {
+          const remoteFilename = `${school._id}_${uuidv4()}_${moment().format('YYYYMMDDHHmmss')}_error.csv`
+          const streamLength = 512 * 1000
+          const csvBlobFile = await new Promise((resolve, reject) => {
+            blobService.createBlockBlobFromText('csvuploads', remoteFilename, cvsStr, streamLength,
+              (error, result) => {
+                if (error) reject(error)
+                else return resolve(result)
+              }
+            )
+          })
+          req.session.csvErrorFile = csvBlobFile.name
+        } catch (err) {
+          return next(err)
+        }
+        // render with errors
+        res.hasError = true
+        getAddMultiplePupils(req, res, next)
+      // Save pupils if validation succeeds
+      } else {
+        let pupilIds = []
+        try {
+          const savedPupils = await Pupil.insertMany(pupils, (err) => {
+            if (err) {
+              throw new Error(err)
+            }
+          })
+          savedPupils.map((p) => pupilIds.push(p._id))
+          req.flash('info', `${pupils.length} new pupils have been added`)
+        } catch (err) {
+          return next(err)
+        }
+        pupilIds = JSON.stringify(pupilIds)
+        res.redirect(`/school/pupil-register/lastName/true?hl=${pupilIds}`)
+      }
+    })
+  stream.pipe(csvStream)
```

Comment:
Well this is a fun function.  Does it work?  Have you got TDD unit tests?

Reply:
tests will accompany each service I will have as part of the refactor

Reply:
You will need to unit test the controller as a part of the refactor.

Reply:
I guess I have to...